### ISSUE TITLE
Gracefully handle flutter navigation

### ISSUE THREAD

**Issue by micimize** [zino-app/graphql-flutter#634]
As noted in #485, all queries are refetched on flutter navigation. This is likely because [all widgets get `didChangeDependencies` called when pushing a route](https://github.com/flutter/flutter/issues/18774#issuecomment-503593201).

#352 could be on the way [possible solution](https://github.com/zino-app/graphql-flutter/pull/352#issuecomment-518030627).

This was only reported with older versions, but I highly doubt it has been solved

**Comment by sterlp** [comment]
I followed a bit of the discussion back to https://github.com/flutter/flutter/issues/11655. I see an additional issue here. For some reason, the "old" pages stay in the layout -- in HTML we would stay invisible in the DOM.

I think this is maybe not really intended, right? As if I pop the "old" page it gets anyway rerendered.

Means I agree with the others, either:

1. Remove the old Widgets, so they will also not rerender on navigation, only if we come back
2. Suppress the updates to invisible "Pages" and keep the logic that we rebuild on return

Does this make any sense?

![image](https://user-images.githubusercontent.com/774984/87294571-3d776f80-c504-11ea-8772-2d8a1a5f41b8.png)

**Comment by micimize** [comment]
@sterlp I guess you're asking about flutter navigation in general, and yeah I'm not entirely sure. I think the navigator maintains state but ignores painting operations, letting them do things in the background. The idea being, they're mounted in the history stack, just hidden. That's why we get the re-querying on `didChangeDependencies`, etc.

As for whether it's intended, I think there is an intent to _allow_ previously rendered pages in the stack to do things in the background, but that the current api causes the accrual of unnoticed background widget activity _by default_, which is basically unexpected behavior.

A bit hard to say for sure though, been a while since I was digging into this stuff. One thing to consider: if you initiate a stateful download on route A, then navigate to route B, would you expect the download to be cancelled due to unmount?

**Comment by sterlp** [comment]
@micimize to your question "_download on route A, then navigate to route B, would you expect the download to be cancelled due to unmount?_"

Short answer: **yes**. Let me explain a bit why I think so:

1. This is the default behavior I would get in any other framework like angular, Android, JSF, HTML, etc:
An Activity is suspended, an angular content view removed.

2. From an architecture view even more, as this widget is just a view. Which from my perspective was closed. This means any "state" it still holds is most likely stale if we return to this view.

_Well I understand the benefit here of course too. As we don't have to restore the view to the state the user left it -- like with an android activity._

3. I want to avoid side effects, especially as in dart we run the UI loop in **one** isolate.
Let's assume we really have a long-running REST request - or even worse (a bit more state of the art) a Web-Socket or MQQT connection.
Here we would still receive updates on a hidden view and consume resources from our event loop. It is quite hard to debug such "hidden" updates, as the framework behaves somehow (at least to me) unsuspected.
What would happen if e.g. the request fails? Will the error message be displayed? This error has most likely nothing to do the with page the user is currently on.

If I get it right, we have now on our own to implement "onPause" and "onResume", using the `WidgetsBinding` right? This feels somehow unnatural to me because it is named _navigate_ which somehow triggers page navigation in my head like in angular or HTML.

Long story short I feel personally this is more unintentionally in the framework right now, leading as a side effect to the issues with `didChangeDependencies`. Basically I think they should be _inactive_ until we return.

![image](https://user-images.githubusercontent.com/774984/87462305-04c4bc80-c610-11ea-9014-01ad71ea0292.png)


P.S.:
Taken (3) into account I am not sure if we might see issues in larger apps considering memory and performance.
Thinking more about (2) a hidden update to the view would bring the widget into a state the user hasn't left it, means we lose the feature "yeah continue to use the widget where I left it"
e.g.: the list view was updated and the element I clicked is no longer in the same position as soon I return, which would be very odd.
